### DISCUSSION TITLE
Is it necessary to set secure header for JavaScript file?

### DISCUSSION THREAD

**Discussion by owem777** [OWASP/www-project-secure-headers#64] [Q&A] [answered]
Hi Leaders,

As we know secure headers that can improve web application security to protect against malicious JavaScript, cross-site scripting, and other common attacks. We have already set secure headers for HTML page, but there are some exceptions such as JavaScript file (ex: xxx.yyy.com/plugin/jquery.js)

Is it necessary to set secure header for JavaScript file?
![image](https://user-images.githubusercontent.com/101683102/158504130-86f25c73-dd3e-4de6-9dcd-5727756de84a.png)

Would you please provide the some suggestion for me?
Thank you very much^^

**Comment by righettod** [comment, marked as answer]
Hi,

Thank you very much for your very interesting question 😃 

For your context (**self hosted static JS script**), I propose to set the following [HTTP response header](https://owasp.org/www-project-secure-headers/#div-headers):
* `Strict-Transport-Security`: To ensure that the script is delivered using a secure channel (HTTPS).
* `X-Content-Type-Options`: To prevent MIME type abuse for IE/Chrome context.
* `Cross-Origin-Embedder-Policy` + `Cross-Origin-Opener-Policy` + `Cross-Origin-Resource-Policy`: To prevent a remote site to load your script and also isolate the browsing context at browser level.

@riramar What do you think about my proposal?

@owem777 I hope this can help you, feel free to indicate to us if it is not the case and then we will deep dive into your context 😃

**Reply by pfreitag** [reply]
The answer also depends on if the `content-type` HTTP response header is properly set to something like `application/javascript` or not.

**Reply by righettod** [reply]
Indeed, thank you very much @pfreitag for this clarification, it is important to ensure that the `content-type` is correctly specified 👍 

More information about:
* [content-type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types).
* [content-type for JavaScript](https://support.maxcdn.com/hc/en-us/articles/360036933471-Picking-the-Right-MIME-Type-for-JavaScript).

Note that [Mozilla seems recommending](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#textjavascript) using `text/javascript` as `content-type` value for JS scripts:

![image](https://user-images.githubusercontent.com/1573775/158659459-fe28fd76-bd22-40f6-97b8-f830a8e891a9.png)

However, [CDN like one for JQuery](https://code.jquery.com/jquery-3.6.0.js) use value `application/javascript`:

![image](https://user-images.githubusercontent.com/1573775/158660287-b1a34320-e77a-43e0-9a7b-d51f745a118a.png)

Same thing regarding [Cloudflare JS CDN](https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js):

![image](https://user-images.githubusercontent.com/1573775/158660869-d50bf6f4-1aaf-48f4-9b3f-61cbc40f8343.png)

**Reply by owem777** [reply]
Hi @righettod  and @pfreitag,

Your suggestion has been received. Thank you very much.

**Comment by riramar** [comment]
That's an interesting question and I think you @righettod are right.
I see if I can research more about it and post here if I find something that make sense.

**Comment by mozfreddyb** [comment]
There have been numerous attacks that relied on abusing non-HTML resources as HTML.
1. [Protecting Browsers from Cross-Origin CSS Attacks](https://www.linshunghuang.com/papers/css.pdf) from 2010
2. [Abusing JSONP with Rosetta Flash](https://blog.miki.it/2014/7/8/abusing-jsonp-with-rosetta-flash/)

It's strongly recommended to set headers like X-Content-Type-Options and Content-Type on those resources. Essentially, the requests to all same-origin/same-site resources are cookied and authenticated. Best to set all headers.

**Reply by righettod** [reply]
Thank you very much for this additional very useful answer 👍

**Reply by owem777** [reply]
Hi Leaders, 

I have another issue about setting secure header for RESTful APIs and need your suggestion, thanks.
I set six secure headers for HTML page (web-based) as below:
No1. Content-Security-Policy
No2. X-Frame-Options
No3. X-XSS-Protection
No4. X-Content-Type-Options
No5. Referrer-Policy 
No6. HTTP Strict-Transport-Security

Are they useful for RESTful APIs?
My understanding: No1~No3 are not necessary due to RESTful APIs can't execute script and frame.
Please correct me if my understanding is not correct.

Thank you very much!

**Reply by righettod** [reply]
Hi @owem777 

Assuming your API only return JSON content type (and without any CORS headers), I will say like you but adding `Content-Security-Policy` to harden the usage context of the response of the API.

@riramar What do you think about it?

**Comment by riramar** [comment]
Without a context it's difficult to determine which headers and values.
Keep in mind all these headers are designed to be consumed by a browser. So if your client it's not a browser there'll be no effect on them.
If there is a malicious scenario where an attack can use an API response to take some advantage those headers can help you to mitigate this scenario.

**Reply by owem777** [reply]
Hi @riramar 

 Thank you for suggestion.

**Comment by righettod** [comment]
@riramar Can you remove the ✔ emoji in the title?

**Comment by riramar** [comment]
Done.
